### ISSUE TITLE
Generate a JWT for activated user

### ISSUE THREAD

**Issue by mislavmatijevic** [foivz/pop-app-backend#9]
# Feature

This feature enables the user to obtain a valid JWT using the right credentials, if activated.
In short, it means it allows a user to log in.

## Description

Consult [FR-3](https://github.com/foivz/pop-app-backend/wiki/SRS#fr-3), [FR-7](https://github.com/foivz/pop-app-backend/wiki/SRS#fr-7) and [defined API endpoints](https://github.com/foivz/pop-app-backend/wiki/SRS#212-endpoints).

| Condition | Description |
| ------ | ---------- |
| Given | User has registered and is activated by an Admin. |
| When | User sends POST _request_ to `/api/v2/auth/login` |
| Then | User receives a _response_ (**if registered in the current event**)
| Or | User receives one of the defined errors, ranging 301-303, or a warning 304
| Or | JWT expires and then user goes to `/api/v2/auth/refresh-token`
| And | User receives a new pair of tokens

Hashing should be performed using MD-5 for salt and SHA-512 for password hash (with salt prefixed to the password).
Both hash and the salt are stored in their respective ["password_salt" and "password_hash" attributes defined in the "user" entity](https://github.com/foivz/pop-app-backend/wiki/ER-Diagram#new-diagram-v1) during registration, so that's where they should be taken from.

## Request for tokens definition

Sample request is made of the following attributes:
| Attribute name | Attribute type | Attribute description |
| --------------- | --------------- | --------------------- |
| username | string | User's username. | 
| password | string | User's password. | 

JSON of the sample request:
```
{
    "username": ihorvat,
    "password": test123
}
```

### Success:
```
{
    "success": true,
    "message": "User \"ihorvat\" logged in",
    "data": [
        {
            "id": 57,
            "first_name": "Ivan",
            "last_name": "Horvat",
            "username": "ihorvat",
            "email": "ihorvat@foi.hr",
            "role": "buyer",
            "activated": true,
            "store": {
                 "store_id": 1,
                 "storename": "SuperStore"
            },
            {
                 "access_token": "eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiIxIiwiZXhwIjoxNjg5MTAwMDAwLCJpYXQiOjE2MzkwMDIxNDF9.ijhNoPEBLHW0PC5fiVYB0xAZ5xj95f-EzZfI-PjtzGU",
                 "refresh_token": {
                      "token": "4VipdFCzDbnse8Lo/k8Pkh04XrW47S623pIRRlcm3XJ5ADBsAtIyyVtWbbrsTWEr",
                      "valid_for": {
                           "time_unit": "minutes",
                           "time_amount": 1440
                      }
                 }
            }
        }
    ]
}
```

### Success, but user has no store:
```
{
    "success": true,
    "message": "User \"ihorvat\" logged in with warnings.",
    "error_code": 304,
    "error_message": "WARN_STORE_NOT_SET",
    "data": [
            {
                "id": 57,
                "first_name": "Ivan",
                "last_name": "Horvat",
                "username": "ihorvat",
                "email": "ihorvat@foi.hr",
                "role": "buyer",
                "activated": true,
                "store": null
            },
            {
                 "access_token": "eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiIxIiwiZXhwIjoxNjg5MTAwMDAwLCJpYXQiOjE2MzkwMDIxNDF9.ijhNoPEBLHW0PC5fiVYB0xAZ5xj95f-EzZfI-PjtzGU",
                 "refresh_token": {
                      "token": "4VipdFCzDbnse8Lo/k8Pkh04XrW47S623pIRRlcm3XJ5ADBsAtIyyVtWbbrsTWEr",
                      "valid_for": {
                           "time_unit": "minutes",
                           "time_amount": 1440
                      }
                 }
            }
    ]
}
```
Note that store attribute does exist, but it is null. Hence the `error_code` and `error_message` attributes, although the status is "true".

Discussion below should decide whether this user even is entitled to a JWT. For now, that detail can be postponed for later decision.
- If JWT is not sent, then there is no safety mechanism to ensure that not anyone can set a store to any user just using their ID - a terrible issue that would have to be taken care of on some other way. However, if no JWT, `error_code` makes sense.
- If JWT is sent, then it is kind of misleading since it still doesn't allow user to do anything other than set a store. However, in this case the `error_code` and `error_message` attributes could confuse the front-end users that the request somehow failed, even though a perfectly valid JWT is returned.

### Failure example:
```
{
    "success": false,
    "message": "Could not log in user \"ihorvat\".",
    "error_code": 303,
    "error_message": "ERR_NOT_ACTIVATED: ihorvat"
}
```

## Refresh token request

Sample request is made of the following attributes:
| Attribute name | Attribute type | Attribute description |
| --------------- | --------------- | --------------------- |
| refresh_token | 64-char string | Last issued refresh token. Becomes invalidated after a successful request. | 

JSON of the sample request:
```
{
    "refresh_token": "4VipdFCzDbnse8Lo/k8Pkh04XrW47S623pIRRlcm3XJ5ADBsAtIyyVtWbbrsTWEr"
}
```

### Success:
```
{
   "access_token": "eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJ0ZXN0ZXJVc2VybmFtZSIsImlhdCI6MTY5MzA4NzE3OSwiZXhwIjoxNjkzMDg4Mzc5fQ.BrFcE9kAWkXGAY_MAfy8dsT3mgpntFfmMkYPUE1Yy80",
   "refresh_token": {
        "token": "5ayrfzCfDbnse8Lo/pIRRlcm3rsTWErk8Pkh04XrW47S623XJ5ADBsAtIyyVtWbb",
        "valid_for": {
             "time_unit": "minutes",
             "time_amount": 1440
        }
   }
}
```

### Failure example:
```
{
    "success": false,
    "message": "Refresh token expired!",
    "error_code": 5,
    "error_message": "ERR_REFRESH_TOKEN_EXPIRED"
}
```

**Comment by mislavmatijevic** [comment]
Also, before implementation - a good discusion would be where to put the JWT itself - as an attribute to the user object (first in data array), as a seperate object in the "data" array, as a separate special attribute in the response itself? ...

**Comment by mislavmatijevic** [comment]
> Also, before implementation - a good discusion would be where to put the JWT itself - as an attribute to the user object (first in data array), as a seperate object in the "data" array, as a separate special attribute in the response itself? ...

I have decided to go with the option where JWT is a second element in the "data" array. It just makes more sense to get JWT independently from User information.

**Comment by mislavmatijevic** [comment]
Hey, @antoniokunstek! Welcome back to Varaždin 😉
I've got a problem, maybe you can help solve it (since I won't be able to deal with the project most probably until Monday).
H2 works fine, but tables return no rows. (note that the rows **are there** - the tests which test the quantity of rows pass just as they did before!)
My presumption is that Spring Security is getting in the way... somewhere? I did spend some fair time trying to figure that out but so far failed.

After this is fixed, I only need to handle JWT expiration. After that, finally, a big nice PR for you. 😃

**Comment by mislavmatijevic** [comment]
> Hey, @antoniokunstek! Welcome back to Varaždin 😉 I've got a problem, maybe you can help solve it (since I won't be able to deal with the project most probably until Monday). H2 works fine, but tables return no rows. (note that the rows **are there** - the tests which test the quantity of rows pass just as they did before!) My presumption is that Spring Security is getting in the way... somewhere? I did spend some fair time trying to figure that out but so far failed.
> 
> After this is fixed, I only need to handle JWT expiration. After that, finally, a big nice PR for you. 😃

It is now fixed by d4078a53.

**Comment by mislavmatijevic** [comment]
I've decided to implement a refresh token, as well.
